### PR TITLE
mintro: Introspection interpreter refactoring

### PR DESCRIPTION
**Pull request by mensinda**
This PR removes some code duplication in mintro, by moving `detect_compilers` completely into `Environment` and by merging both introspection interpreters.

cc: @textshell, @Ericson2314

### CODE REVIEW

File: mesonbuild/environment.py [hunk 1]
```diff
@@ -967,7 +967,7 @@ def detect_swift_compiler(self):
             return compilers.SwiftCompiler(exelist, version)
         raise EnvironmentException('Unknown compiler "' + ' '.join(exelist) + '"')
 
-    def detect_compilers(self, lang, need_cross_compiler):
+    def detect_compilers(self, lang: str, need_cross_compiler: bool, introspection_mode: bool = False):
         comp = None
         cross_comp = None
```
Comment (Ericson2314) [inline mesonbuild/environment.py]: I don't think you need these anymore...

File: mesonbuild/environment.py [hunk 2]
```diff
@@ -1016,6 +1016,37 @@ def detect_compilers(self, lang, need_cross_compiler):
                 raise EnvironmentException('Cross compilation with Swift is not working yet.')
                 # cross_comp = self.environment.detect_fortran_compiler(True)
 
+        if comp is None:
```
Comment (Ericson2314) [inline mesonbuild/environment.py]: ...if you make this a final `else` and the rest `elif`s

File: mesonbuild/environment.py [hunk 3]
```diff
@@ -1016,6 +1016,37 @@ def detect_compilers(self, lang, need_cross_compiler):
...
+            if not introspection_mode:
+                raise EnvironmentException('Tried to use unknown language "%s".' % lang)
+            else:
+                return None, None
+
```
Comment (Ericson2314) [inline mesonbuild/environment.py]: If we split right here to make a new function, the caller can

1. Call the old function, which is what was already in `environment.py`
2. Do the abort themselves, as before
3. Do the `sanity_check` calls, if applicable
4. Call the new function

File: mesonbuild/environment.py (hunk 2 above)
Comment (mensinda) [reply mesonbuild/environment.py]: Obsolete with the last commit.

File: mesonbuild/environment.py (hunk 3 above)
Comment (mensinda) [reply mesonbuild/environment.py]: Sure, but I split it a bit differently: `detect_compilers` now calls both the old and the new function. The user can then decide to call the additional `check_compilers` function if necessary. This approach reduces the amount of code duplication and extra logic in the interpreter and introspector.

Comment (Ericson2314) [reply mesonbuild/environment.py]: Even better!

**Comment by jpakkane** [comment]
Looks good, but since we are touching this, the class name should be `IntrospectionInterpreter` rather than `IntrospectionInterperter`.

**Comment by mensinda** [comment]
Oh... I missed that. Fixed the spelling.

**Comment by jpakkane** [comment]
Please rebase this so that the last commit is merged to the first one. With that fixed it's ok to merge.

**Comment by mensinda** [comment]
@jpakkane Rebased.

**Comment by mensinda** [comment]
The CI error seems to be unrelated.

**Comment by Ericson2314** [comment]
Hmm upon further thought, I think `detect_compilers` and `process_new_compilers` should go in `coredata`. The actual detecting of compilers already is in `environment`; this code really is just dispatching to the right one and handling various options, all of which are an coredata concern. `environment` should be about detecting meson-agnostic things in the larger world.

**Comment by mensinda** [comment]
My problem with moving `process_new_compilers` into `coredata` is that this would require passing `environment` to `process_new_compilers`, since `environment.cmd_line_options` are used there. In my opinion, this is worse than leaving it in `environment`.

Alternatively `cmd_line_options` could also be moved to `coredata`, but I don't know if that breaks the serialization or makes sense.

**Comment by Ericson2314** [comment]
I think just passing in `cmd_line_options` is options would be fine?

**Comment by mensinda** [comment]
My problem with passing `cmd_line_options` is that it is not obvious (when only looking at the code in `coredata`) where these `cmd_line_options` are coming from. You would have to find the two instances where `process_new_compilers` is used to know how to use it.

That being said, `cmd_line_options` only seem to be used in `process_new_compilers` (aside from `__init__`), so moving them into `coredata` with the function itself should be fine.

@jpakkane Any opinions regarding this? Can `cmd_line_options` be safely moved into `coredata` without breaking serialization, etc.?

**Comment by Ericson2314** [comment]
Also, to be clear, I have 0 problem with this all happening in a follow-up PR :).

**Comment by mensinda** [comment]
OK. Moved `process_new_compilers` to `coredata` with the `cmd_line_options`.

**Comment by Ericson2314** [comment]
Great, thanks!

**Comment by jpakkane** [comment]
Ping @dcbaker as per above.

**Comment by dcbaker** [comment]
@jpakkane, still happy to merge this?

**Comment by Ericson2314** [comment]
I opened https://github.com/mesonbuild/meson/pull/4781 to move the field we don't need/want to persist back to `environment`.

CC @textshell

**Comment by mensinda** [comment]
@Ericson2314 Merged.

**Comment by dcbaker** [comment]
with the cmd_line pulled back into environment there shouldn't be any problems with my work.

**Comment by Ericson2314** [comment]
Thanks!!

**Comment by jpakkane** [comment]
:crossed_swords: conflicts...

**Comment by mensinda** [comment]
And rebased

**Comment by Ericson2314** [comment]
@mensinda you lost my commit on the rebase I think? (it doesn't look squashed away either).

**Comment by Ericson2314** [comment]
@mensinda I'm dredging it up in case that helps.

**Comment by Ericson2314** [comment]
@mensinda To be honest I would just squash it into the commit that moved the field in the first place so as to avoid the there-and-back churn.

**Comment by mensinda** [comment]
And it the commit is back in again.

**Comment by Ericson2314** [comment]
I just force-pushed my branch after your merge with that that squash.

**Comment by Ericson2314** [comment]
To be clear, the final state of the PR is far more important than the history, so this is still good to go as far as I'm concerned. :crossed_fingers: no conflicts this time!

**Pull request merged**